# Steam games
###### Proceso de ETL del archivo `steam_games.json.gz`

### Librerías a usar y definición de funciones

In [1]:
import warnings
warnings.filterwarnings("ignore")

In [2]:
import pandas as pd
import numpy as np
import json
import gzip
import os

In [3]:
pd.options.display.max_columns = None
pd.set_option('display.expand_frame_repr', False)
pd.set_option('display.max_rows', None)
pd.set_option('display.width', 999) 

In [4]:
project_path = os.path.dirname(os.path.realpath('../__file__'))

## Extración e inspección

### Lectura de los datos

Para ver como estaba constituido el archivo, se descomprimió y se abrió en un archivo texto, donde se pudo verificar, que cada línea del archivo constituía un dato almacenado en formato ".JSON"

In [5]:
# Ruta del archivo steam_games.json.gz 
file_path = os.path.join(project_path, 'Data\source_DATA\steam_games.json.gz')

# Lista para almacenar los datos
data_list = []

# Descomprime el archivo y lee línea por línea
with gzip.open(file_path, 'rt', encoding='utf-8') as file:
    
    for line in file:
        
        # Carga la línea como JSON
        data = json.loads(line)
        data_list.append(data)

# Crea un DataFrame a partir de la lista de datos
df_games = pd.DataFrame(data_list)

### Conociendo al Dataset

En este momento, usaremos los distintos atributos de los DataGrames de Pandas, para tener una idea de como están conformados los datos y de cuantos hay

In [6]:
df_games.head(20)

publisher genres app_name title  url release_date tags reviews_url specs price early_access   id developer
0        NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
1        NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
2        NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
3        NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
4        NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
5        NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
6        NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
7        NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
8        NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
9        NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
10       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
11       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
12       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
13       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
14       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
15       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
16       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
17       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
18       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN
19       NaN    NaN      NaN   NaN  NaN          NaN  NaN         NaN   NaN   NaN          NaN  NaN       NaN

In [7]:
df_games.tail(20)

publisher                                             genres                                         app_name                                            title                                                url  release_date                                               tags                                        reviews_url                                              specs  price early_access      id                                  developer
120425  Spiderweb Software                             [Strategy, RPG, Indie]                                      Geneforge 2                                      Geneforge 2  http://store.steampowered.com/app/200980/Genef...    2003-07-01  [RPG, Indie, Strategy, Turn-Based Combat, Isom...  http://steamcommunity.com/app/200980/reviews/?...                                    [Single-player]  19.99        False  200980                         Spiderweb Software
120426    Epic Games, Inc.                                           [Action]  Unreal Tournament 2004: Editor's Choice Edition  Unreal Tournament 2004: Editor's Choice Edition  http://store.steampowered.com/app/13230/Unreal...    2004-03-16  [FPS, Arena Shooter, Action, Classic, Multipla...  http://steamcommunity.com/app/13230/reviews/?b...  [Single-player, Multi-player, Steam Trading Ca...  14.99        False   13230                           Epic Games, Inc.
120427               Valve                                           [Action]                   Counter-Strike: Condition Zero                   Counter-Strike: Condition Zero  http://store.steampowered.com/app/80/CounterSt...    2004-03-01  [Action, FPS, Shooter, Multiplayer, Singleplay...  http://steamcommunity.com/app/80/reviews/?brow...  [Single-player, Multi-player, Valve Anti-Cheat...   9.99        False      80                                      Valve
120428                 NaN                                                NaN        Robotpencil Presents: Exercise: Brushwork        Robotpencil Presents: Exercise: Brushwork  http://store.steampowered.com/app/775640/Robot...    2018-01-03                  [Design & Illustration, Tutorial]  http://steamcommunity.com/app/775640/reviews/?...                                                NaN   3.99        False  775640                                        NaN
120429                 NaN                                                NaN       Robotpencil Presents: Creative Composition       Robotpencil Presents: Creative Composition  http://store.steampowered.com/app/777930/Robot...    2018-01-03                  [Design & Illustration, Tutorial]  http://steamcommunity.com/app/777930/reviews/?...                                                NaN   3.99        False  777930                                        NaN
120430                 NaN                                                NaN                                 The Gamble House                                 The Gamble House  http://store.steampowered.com/app/775370/The_G...    2016-11-19                                            [Movie]  http://steamcommunity.com/app/775370/reviews/?...                               [Captions available]   4.99        False  775370                                        NaN
120431                 NaN                                                NaN         Kalen Chock Presents: 2017 Free Tutorial         Kalen Chock Presents: 2017 Free Tutorial  http://store.steampowered.com/app/777950/Kalen...    2018-01-03                  [Design & Illustration, Tutorial]  http://steamcommunity.com/app/777950/reviews/?...                                                NaN   Free        False  777950                                        NaN
120432           Fred Ells                                            [Indie]                          Agent X: Equation Rider                          Agent X: Equation Rider  http://store.steampowered.com/app/767010/Agent...    2018-01-03                                            [Indie]  http

In [8]:
df_games.shape

(120445, 13)

Primeramente se observan gran cantidad de datos nulos los cuales no aportan ninguna información. Luego se aprecian que las columnas `genres` `tags` y `specs` existen datos en anidados en forma de lista, y por último, que el el dataframe consta de 120445 datos y 13 atributos.

## Transformación

### Tratamiento de datos nulos

In [9]:
# Eliminación de datos nulos
df_games = df_games.dropna(how='all').reset_index(drop=True)
df_games.shape

(32135, 13)

In [10]:
df_games.isnull().sum()

publisher       8052
genres          3283
app_name           2
title           2050
url                0
release_date    2067
tags             163
reviews_url        2
specs            670
price           1377
early_access       0
id                 2
developer       3299
dtype: int64

### Datos duplicados

Inicialmente usaremos la columna id para ver si hay registros repetidos

In [11]:
duplicated_games_id = df_games.loc[df_games['id'].duplicated(keep=False)]
duplicated_games_id

publisher               genres                                        app_name                                           title                                                url release_date                                               tags                                        reviews_url                                              specs  price early_access      id                                   developer
74                                                   NaN                  NaN                                             NaN                                             NaN                     http://store.steampowered.com/          NaN                                                NaN                                                NaN                                                NaN  19.99        False     NaN                                         NaN
13894                                 Bethesda Softworks             [Action]                Wolfenstein II: The New Colossus                Wolfenstein II: The New Colossus          http://store.steampowered.com/app/612880/   2017-10-26  [Action, FPS, Gore, Violent, Alternate History...  http://steamcommunity.com/app/612880/reviews/?...  [Single-player, Steam Achievements, Full contr...  59.99        False  612880                               Machine Games
14573                                 Bethesda Softworks             [Action]                Wolfenstein II: The New Colossus                Wolfenstein II: The New Colossus  http://store.steampowered.com/app/612880/Wolfe...   2017-10-26  [Action, FPS, Gore, Violent, Alternate History...  http://steamcommunity.com/app/612880/reviews/?...  [Single-player, Steam Achievements, Full contr...  59.99        False  612880                               Machine Games
30961  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]  Batman: Arkham City - Game of the Year Edition  Batman: Arkham City - Game of the Year Edition           http://store.steampowered.com/app/200260   2012-09-07  [Action, Open World, Batman, Adventure, Stealt...                                                NaN  [Single-player, Steam Achievements, Steam Trad...  19.99        False     NaN  Rocksteady Studios,Feral Interactive (Mac)

Se consiguieron dos `id` repetidos, que son el ***612880*** y otro que toma por valor ***NaN***. Acontinuación contrastaremos los resultados usando la columna `devoloper`.

In [12]:
df_games[df_games['developer']=='Rocksteady Studios,Feral Interactive (Mac)']

publisher               genres                                        app_name                                           title                                                url release_date                                               tags                                        reviews_url                                              specs  price early_access      id                                   developer
1068   Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]  Batman: Arkham City - Game of the Year Edition  Batman: Arkham City - Game of the Year Edition  http://store.steampowered.com/app/200260/Batma...   2012-09-07  [Action, Open World, Batman, Adventure, Stealt...  http://steamcommunity.com/app/200260/reviews/?...  [Single-player, Steam Achievements, Steam Trad...  19.99        False  200260  Rocksteady Studios,Feral Interactive (Mac)
30961  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]  Batman: Arkham City - Game of the Year Edition  Batman: Arkham City - Game of the Year Edition           http://store.steampowered.com/app/200260   2012-09-07  [Action, Open World, Batman, Adventure, Stealt...                                                NaN  [Single-player, Steam Achievements, Steam Trad...  19.99        False     NaN  Rocksteady Studios,Feral Interactive (Mac)
31617  Warner Bros. Interactive Entertainment, Feral ...  [Action, Adventure]  Batman: Arkham Asylum Game of the Year Edition  Batman: Arkham Asylum Game of the Year Edition  http://store.steampowered.com/app/35140/Batman...   2010-03-26  [Action, Batman, Stealth, Adventure, Third Per...  http://steamcommunity.com/app/35140/reviews/?b...  [Single-player, Steam Achievements, Steam Trad...  19.99        False   35140  Rocksteady Studios,Feral Interactive (Mac)

Vemos que el juego `Batman: Arkham City - Game of the Year Edition` aparece dos veces, una vez con el `id` ***200260*** y otra sin `id`. y Porcedemos a eliminar los valores repetidos con índices **74**, **14573**, **30961**


In [13]:
df_games = df_games.drop([ 74, 14573, 30961])

### Simplificación del DATASET


In [14]:
df_games = df_games.drop(['tags', 'specs', 'url', 'reviews_url'], axis=1)

### Datos Nulos


In [15]:
df_games['publisher'].fillna('NO DISPONOBLE', inplace=True)
df_games['app_name'].fillna('NO DISPONOBLE', inplace=True)
df_games['title'].fillna('NO DISPONOBLE', inplace=True)
df_games['developer'].fillna('NO DISPONOBLE', inplace=True)

### Transformación de la Columna 'release_date'


In [16]:
# Convertir la columna 'release_date' al tipo datetime
df_games['release_date'] = pd.to_datetime(df_games['release_date'], errors='coerce')

# Extraer el año y crear la columna 'release_year', asignando -1 a los valores no válidos
df_games['release_year'] = df_games['release_date'].dt.year.fillna(-1).astype(int)

# Reemplazar los valores -1 por "Dato no disponible" en la columna 'release_year'
df_games['release_year'] = df_games['release_year'].replace(-1, "NO DISPONIBLE")

# Eliminar la columna 'release_date' original
df_games.drop('release_date', axis=1, inplace=True)

### Transformación de la Columna 'price'

Existen valores Nulos en la columna `price`, sin embargo son distintos a los valores de gratis o de promoción, que están en `str`. Estos valores gratis y de promoción se les asignará el valor **0**, y los valores nulos se dejaran por defecto ya que son tomados del tipo `float`.

In [17]:
df_games['price'] = df_games['price'].apply(lambda x: 0 if isinstance(x, str) else round(float(x), 2))

### Transformación de la Columna 'genres'


In [18]:
df_games = df_games.explode('genres')

df_games = df_games.dropna(subset=['genres'])

### Verificación de los datos

In [19]:
df_games.describe()

price
count  68352.000000
mean       8.839481
std       16.024124
min        0.000000
25%        2.990000
50%        4.990000
75%        9.990000
max      995.000000

In [20]:
df_games.isnull().sum()

publisher          0
genres             0
app_name           0
title              0
price           3199
early_access       0
id                 0
developer          0
release_year       0
dtype: int64

In [21]:
df_games.shape

(71551, 9)

## Carga

In [22]:
# Método de compresión
compression_opts = dict(method='gzip')

# Creación del archivo
df_games.to_csv(os.path.join(project_path, 'Data\clean_DATA\SteamGames.csv.gz'), index=False, encoding='utf-8', compression = compression_opts)

Se creó un archivo `.csv` comprimido con el nombre *`SteamGames.csv.gz`*, en la subcarpeta `\clean_Data` de la carpeta `\Data`.